File: src/main/java/com/sosnoski/concur/article5java8/Hello2.java
package com.sosnoski.concur.article5java8;

import akka.actor.*;
import akka.japi.pf.ReceiveBuilder;

/** Simple hello from an actor with state. */
public class Hello2 {

    public static void main(String[] args) {
        ActorSystem system = ActorSystem.create("actor-demo-java8");
        ActorRef hello = system.actorOf(Props.create(Hello.class));
        hello.tell(new Greeting("Hello"), ActorRef.noSender());
        hello.tell(new Greet("Bob"), ActorRef.noSender());
        hello.tell(new Greet("Alice"), ActorRef.noSender());
        hello.tell(new Greeting("Hola"), ActorRef.noSender());
        hello.tell(new Greet("Alice"), ActorRef.noSender());
        hello.tell(new Greet("Bob"), ActorRef.noSender());
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) { /* ignore */ }
        system.shutdown();
    }
    
    private static class Greeting {
        public final String greeting;
        
        public Greeting(String text) {
            greeting = text;
        }
    }
    
    private static class Greet {
        public final String name;
        
        public Greet(String text) {
            name = text;
        }
    }

    private static class Hello extends AbstractActor {
        private String greeting;
        
        public Hello() {
            receive(ReceiveBuilder.
                match(Greeting.class, g -> { greeting = g.greeting; }).
                match(Greet.class, g -> { System.out.println(greeting + " " + g.name); }).
                build());
        }
    }
}

File: src/main/scala/com/sosnoski/concur/article5scala/Hello2.scala
package com.sosnoski.concur.article5scala

import akka.actor._
import akka.util._

/** Simple hello from an actor with state. */
object Hello2 extends App {
  
  case class Greeting(greet: String)
  case class Greet(name: String)
  
  val system = ActorSystem("actor-demo-scala")
  val hello = system.actorOf(Props[Hello])
  hello ! Greeting("Hello")
  hello ! Greet("Bob")
  hello ! Greet("Alice")
  hello ! Greeting("Hola")
  hello ! Greet("Alice")
  hello ! Greet("Bob")
  Thread sleep 1000
  system shutdown
  
  class Hello extends Actor {
    var greeting = ""
    def receive = {
      case Greeting(greet) => greeting = greet
      case Greet(name) => println(s"$greeting $name")
    }
  }
}